pkg/driver/controller.go

```diff
@@ -101,107 +101,122 @@ func getRoundedCapacity(size int64) int64 {
 	return ((size + Mi - 1) / Mi) * Mi
 }
 
-func waitForReadyVolume(volname string) error {
-	for true {
-		vol, err := lvm.GetLVMVolume(volname)
-		if err != nil {
-			return status.Errorf(codes.Internal,
-				"lvm: wait failed, not able to get the volume %s %s", volname, err.Error())
+// waitForLVMVolume waits for completion of any processing of lvm volume.
+// It returns the final status of lvm volume along with a boolean denoting
+// whether it should be rescheduled on some other volume group or node.
+// In case volume ends up in failed state and rescheduling is required,
+// func is also deleting the lvm volume resource, so that it can be
+// re provisioned on some other node.
+func waitForLVMVolume(ctx context.Context,
+	vol *lvmapi.LVMVolume) (*lvmapi.LVMVolume, bool, error) {
+	var reschedule bool // tracks if rescheduling is required or not.
+	var err error
+	if vol.Status.State == lvm.LVMStatusPending {
+		if vol, err = lvm.WaitForLVMVolumeProcessed(ctx, vol.GetName()); err != nil {
+			return nil, false, err
 		}
+	}
+	// if lvm volume is ready, return the provisioned node.
+	if vol.Status.State == lvm.LVMStatusReady {
+		return vol, false, nil
+	}
 
-		switch vol.Status.State {
-		case lvm.LVMStatusReady:
-			return nil
-		}
-		time.Sleep(time.Second)
+	// Now it must be in failed state if not above. See if we need
+	// to reschedule the lvm volume.
+	var errMsg string
+	if volErr := vol.Status.Error; volErr != nil {
+		errMsg = volErr.Message
+		reschedule = true
+	} else {
+		errMsg = fmt.Sprintf("failed lvmvol must have error set")
 	}
-	return nil
-}
 
-func waitForVolDestroy(volname string) error {
-	for true {
-		_, err := lvm.GetLVMVolume(volname)
-		if err != nil {
-			if k8serror.IsNotFound(err) {
-				return nil
-			}
-			return status.Errorf(codes.Internal,
-				"lvm: destroy wait failed, not able to get the volume %s %s", volname, err.Error())
+	if reschedule {
```

Comment:
We can since reschedule var is only set to true at one place . Just keeping it here for readability purpose where we first take decision on whether we want to reschedule or not and then at last we run some clean up actions based on if rescheduling is required. I can move it if you think otherwise.